src/dataunion/DataUnion.ts

```diff
@@ -564,30 +590,109 @@ export class DataUnion {
+
+        log('Executing DataUnionSidechain withdraw function')
         const tx = await getWithdrawTxFunc()
         const tr = await tx.wait()
-        if (!freeWithdraw && sendToMainnet) {
-            await this.getContracts().transportSignaturesForTransaction(tr, options)
+
+        // keep tokens in the sidechain => just return the sidechain tx receipt
+        if (!sendToMainnet) { return tr }
+
+        log(`Got receipt, filtering UserRequestForSignature from ${tr.events!.length} events...`)
+        const ambHashes = getMessageHashes(tr)
+
+        if (ambHashes.length < 1) {
+            throw new Error("No UserRequestForSignature events emitted from withdraw transaction, can't transport withdraw to mainnet")
+        }
+
+        if (ambHashes.length > 1) {
+            throw new Error('Expected only one UserRequestForSignature event')
+        }
+
+        const messageHash = ambHashes[0]
+
+        if (!payForTransport) {
+            // expect someone else to do the transport for us (corresponds to dataUnion.freeWithdraw=true)
+            if (waitUntilTransportIsComplete) {
+                log(`Waiting for balance=${balanceBefore.toString()} change (poll every ${pollingIntervalMs}ms, timeout after ${retryTimeoutMs}ms)`)
+                await until(async () => !(await getBalanceFunc()).eq(balanceBefore), retryTimeoutMs, pollingIntervalMs).catch((e) => {
+                    const msg = `Timeout: Bridge did not transport withdraw message as expected. Fix: DataUnion.transportMessage("${messageHash}")`
+                    throw e.message.startsWith('Timeout') ? new Error(msg) : e
+                })
+                return null
+            }
+
+            // instead of waiting, hand out the messageHash so that someone else might do the transport using it
+            return messageHash
+        }
+
+        const ambTr = await this.transportMessage(messageHash, pollingIntervalMs, retryTimeoutMs)
+        if (waitUntilTransportIsComplete) {
+            log(`Waiting for balance ${balanceBefore.toString()} to change`)
+            await until(async () => !(await getBalanceFunc()).eq(balanceBefore), retryTimeoutMs, pollingIntervalMs)
+        }
+        return ambTr
+    }
+
+    // TODO: this doesn't belong here. Transporting a message is NOT dataunion-specific and needs nothing from DataUnion.ts.
+    //       It shouldn't be required to create a DataUnion object to call this.
+    //       This belongs to the StreamrClient, and if the code is too DU-specific then please shove it back to Contracts.ts.
+    //       Division to transportMessage and Contracts.transportSignaturesForMessage is spurious, they should be one long function probably.
+    /**
+     * @returns null if message was already transported, ELSE the mainnet AMB signature execution transaction receipt
+     */
+    async transportMessage(messageHash: AmbMessageHash, pollingIntervalMs: number = 1000, retryTimeoutMs: number = 300000) {
```

Comment:
The blockchain-level code (`transportMessage` and part of `_executeWithdraw`) could be moved to `dataunion/Contracts.ts`

Reply:
Made a ticket https://linear.app/streamr/issue/ETH-101/streamr-client-javascript-data-union-ethereum-utils-refactor